File: backend/order_service/app/schemas/order.py
from pydantic import BaseModel, Field, validator
from typing import List, Optional
from datetime import datetime
from enum import Enum
import json

class OrderStatus(str, Enum):
    PENDING = "pending"
    CONFIRMED = "confirmed"
    PREPARING = "preparing"
    OUT_FOR_DELIVERY = "out_for_delivery"
    DELIVERED = "delivered"
    CANCELLED = "cancelled"
    RETURN_REQUESTED = "return_requested"
    RETURN_APPROVED = "return_approved"
    PICKED_UP = "picked_up"
    REFUNDED = "refunded"

class ProductResponse(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    price: float
    mrp: Optional[float] = None
    category_id: Optional[int] = None
    image_url: Optional[str] = None
    stock_quantity: Optional[int] = None
    unit: Optional[str] = None
    is_active: Optional[bool] = None
    created_at: Optional[datetime] = None

class OrderItemResponse(BaseModel):
    id: int
    product_id: int
    quantity: int
    price: float
    product: Optional[ProductResponse] = None
    
    class Config:
        from_attributes = True

class OrderItemCreate(BaseModel):
    product_id: int
    quantity: int = Field(..., gt=0)

class OrderCreate(BaseModel):
    delivery_address: str
    delivery_latitude: Optional[str] = None
    delivery_longitude: Optional[str] = None
    scheduled_for: Optional[datetime] = None

class OrderUpdate(BaseModel):
    delivery_address: Optional[str] = None
    delivery_latitude: Optional[str] = None
    delivery_longitude: Optional[str] = None

class OrderResponse(BaseModel):
    id: int
    user_id: str
    total_amount: float
    delivery_fee: float
    status: OrderStatus
    delivery_address: str
    delivery_latitude: Optional[str] = None
    delivery_longitude: Optional[str] = None
    estimated_delivery_time: Optional[datetime] = None
    delivered_at: Optional[datetime] = None
    cancelled_at: Optional[datetime] = None
    scheduled_for: Optional[datetime] = None
    created_at: datetime
    items: List[OrderItemResponse] = []
    
    class Config:
        from_attributes = True

class OrderStatusUpdate(BaseModel):
    status: OrderStatus

class AssignDeliveryPartnerRequest(BaseModel):
    delivery_partner_id: str

class OrderCancelResponse(BaseModel):
    success: bool
    message: str
    refund_initiated: bool

class OrderItemsUpdate(BaseModel):
    items: List[OrderItemCreate]

class OrderTemplateCreate(BaseModel):
    name: str
    items: List[OrderItemCreate]

class OrderTemplateResponse(BaseModel):
    id: int
    user_id: str
    name: str
    items: List[OrderItemCreate]
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

class OrderFeedbackCreate(BaseModel):
    rating: int = Field(..., ge=1, le=5)
    comment: Optional[str] = None

class OrderFeedbackResponse(BaseModel):
    id: int
    order_id: int
    rating: Optional[int] = None
    comment: Optional[str] = None
    created_at: datetime
    
    class Config:
        from_attributes = True

class BulkStatusUpdate(BaseModel):
    order_ids: List[int]
    status: OrderStatus

class BulkAssignDelivery(BaseModel):
    order_ids: List[int]
    delivery_partner_id: str

class OrderAnalyticsResponse(BaseModel):
    date: datetime
    order_count: int
    total_revenue: float
    average_delivery_time: Optional[float] = None

class CancellationRateResponse(BaseModel):
    period: str
    cancellation_rate: float
    total_orders: int
    cancelled_orders: int

class TopCustomerResponse(BaseModel):
    user_id: str
    order_count: int
    total_spent: float

File: frontend/src/types/order.ts
// Order types based on the backend order service specification

export enum OrderStatus {
    PENDING = "pending",
    CONFIRMED = "confirmed",
    PREPARING = "preparing",
    OUT_FOR_DELIVERY = "out_for_delivery",
    DELIVERED = "delivered",
    CANCELLED = "cancelled",
    RETURN_REQUESTED = "return_requested",
    RETURN_APPROVED = "return_approved",
    PICKED_UP = "picked_up",
    REFUNDED = "refunded"
}

export interface Product {
    id: number;
    name: string;
    description: string;
    price: number;
    mrp: number;
    category_id: number;
    image_url: string;
    stock_quantity: number;
    unit: string;
    is_active: boolean;
    created_at: string;
}

export interface OrderItem {
    id: number;
    product_id: number;
    quantity: number;
    price: number;
    product?: Product;
}

export interface Order {
    id: number;
    user_id: string;
    delivery_partner_id?: string;
    total_amount: number;
    delivery_fee: number;
    status: OrderStatus;
    delivery_address: string;
    delivery_latitude?: string;
    delivery_longitude?: string;
    estimated_delivery_time?: string;
    delivered_at?: string;
    cancelled_at?: string;
    scheduled_for?: string;
    created_at: string;
    updated_at?: string;
    items: OrderItem[];
}

export interface OrderCreate {
    delivery_address: string;
    delivery_latitude?: string;
    delivery_longitude?: string;
    scheduled_for?: string;
}

export interface OrderStatusUpdate {
    status: OrderStatus;
}

export interface AssignDeliveryPartnerRequest {
    delivery_partner_id: string;
}

export interface OrderCancelResponse {
    success: boolean;
    message: string;
    refund_initiated: boolean;
}

export interface OrderItemsUpdate {
    items: Array<{
        product_id: number;
        quantity: number;
    }>;
}

export interface OrderTemplate {
    id: number;
    user_id: string;
    name: string;
    items: Array<{
        product_id: number;
        quantity: number;
    }>;
    created_at: string;
    updated_at: string;
}

export interface OrderTemplateCreate {
    name: string;
    items: Array<{
        product_id: number;
        quantity: number;
    }>;
}

export interface BulkStatusUpdate {
    order_ids: number[];
    status: OrderStatus;
}

export interface BulkAssignDelivery {
    order_ids: number[];
    delivery_partner_id: string;
}

export interface OrderFeedbackCreate {
    rating: number;
    comment?: string;
}

export interface OrderFeedbackResponse {
    id: number;
    order_id: number;
    rating: number;
    comment?: string;
    created_at: string;
}

export interface DeliveryLocation {
    order_id: number;
    latitude: string;
    longitude: string;
    last_updated: string;
}

export interface OrderAnalytics {
    revenue: Array<{
        date: string;
        total_revenue: number;
        order_count: number;
        average_delivery_time?: number;
    }>;
    delivery_performance: {
        average_delivery_time: number;
        on_time_delivery_rate: number;
        total_deliveries: number;
    };
    top_customers: Array<{
        user_id: string;
        order_count: number;
        total_spent: number;
    }>;
    cancellation_rate: {
        period: string;
        cancellation_rate: number;
        total_orders: number;
        cancelled_orders: number;
    };
}